toolchain/check/impl.cpp

```diff
@@ -227,17 +232,197 @@ auto ImplWitnessForDeclaration(Context& context, const SemIR::Impl& impl)
+    if (auto decl =
+            context.insts().TryGetAs<SemIR::AssociatedConstantDecl>(decl_id)) {
+      auto& witness_value = witness_block[index];
+      if (witness_value == SemIR::ErrorInst::SingletonInstId) {
+        continue;
+      }
+      auto rewrite_value = rewrite_values[index];
+      if (witness_value.is_valid()) {
+        if (!rewrite_value.is_valid()) {
+          CARBON_DIAGNOSTIC(AssociatedConstantMissingInRedecl, Error,
+                            "associated constant {0} given value in "
+                            "declaration but not redeclaration",
+                            SemIR::NameId);
+          auto builder = context.emitter().Build(
+              impl.latest_decl_id(), AssociatedConstantMissingInRedecl,
+              decl->name_id);
+          NotePreviousDecl(context, builder, prev_decl_id);
+          builder.Emit();
+          continue;
+        }
+        auto witness_const_id = context.constant_values().Get(witness_value);
+        if (witness_const_id != rewrite_value) {
+          // TODO: Figure out how to print the two different values
+          CARBON_DIAGNOSTIC(
+              AssociatedConstantDifferentInRedecl, Error,
+              "redeclaration with different value for associated constant {0}",
+              SemIR::NameId);
+          auto builder = context.emitter().Build(
+              impl.latest_decl_id(), AssociatedConstantDifferentInRedecl,
+              decl->name_id);
+          NotePreviousDecl(context, builder, prev_decl_id);
+          builder.Emit();
+          continue;
+        }
+      } else if (rewrite_value.is_valid()) {
+        witness_value = context.constant_values().GetInstId(rewrite_value);
+      }
+    }
+  }
+}
+
+auto ImplWitnessStartDefinition(Context& context, SemIR::Impl& impl) -> void {
   CARBON_CHECK(impl.is_being_defined());
-  // TODO: Check we have a value for all non-function associated constants in
-  // the constraint, and fill the witness with them.
+  CARBON_CHECK(impl.witness_id.is_valid());
+  if (impl.witness_id == SemIR::ErrorInst::SingletonInstId) {
+    return;
+  }
+
+  auto facet_type_id = context.GetTypeIdForTypeInst(impl.constraint_id);
+  CARBON_CHECK(facet_type_id != SemIR::ErrorInst::SingletonTypeId);
+  auto facet_type = context.types().GetAs<SemIR::FacetType>(facet_type_id);
+  const SemIR::FacetTypeInfo& facet_type_info =
+      context.facet_types().Get(facet_type.facet_type_id);
+
+  auto interface_type = facet_type_info.TryAsSingleInterface();
+  CARBON_CHECK(interface_type.has_value());
+  const auto& interface =
+      context.interfaces().Get(interface_type->interface_id);
+
+  auto witness = context.insts().GetAs<SemIR::ImplWitness>(impl.witness_id);
+  auto witness_block = context.inst_blocks().GetMutable(witness.elements_id);
+  auto assoc_entities =
+      context.inst_blocks().Get(interface.associated_entities_id);
+  CARBON_CHECK(witness_block.size() == assoc_entities.size());
```

Comment:
We seem to be doing these same setup steps to prepare to mutate a witness table in three places (`AddConstantsToImplWitnessFromConstraint`, `ImplWitnessStartDefinition`, and `FinishImplWitness`). I wonder if it'd be worth factoring this out somehow -- eg, moving these variables into a struct (and maybe making the functions that operate on them member functions).

Reply:
(It's fine to defer this to a later PR or just add a TODO.)